### ISSUE TITLE
Add remap3d dense volumetric resampling primitive

### ISSUE THREAD

**Issue by ternaus** [albumentations-team/albucore#154]
## Goal

Add `remap3d`, a low-level primitive that applies one caller-provided dense 3D pull-sampling grid to one CPU volume. AlbumentationsX needs this primitive for [`ElasticTransform3D`](https://github.com/albumentations-team/AlbumentationsX/issues/327): the transform will build one deformation grid and reuse it for the volume and aligned masks.

Albucore already has `warp_affine3d`, which generates an affine grid and samples through PyTorch's 5D CPU `grid_sample`. `remap3d` should reuse only the dtype, fill, container, and sampling machinery that remains fast for both primitives. Affine matrix inversion and affine-grid generation stay in `warp_affine3d`.

## Proposed contract

```python
remap3d(
    volume,
    sampling_grid,
    interpolation=cv2.INTER_LINEAR,
    border_mode=cv2.BORDER_CONSTANT,
    border_value=None,
)
```

- `volume` is a NumPy `DHWC` array or, if the direct Tensor route passes its benchmark gate, a CPU Torch `CDHW` tensor.
- Supported volume dtypes are `uint8` and `float32`.
- `sampling_grid` is a CPU float32 NumPy array or Tensor with shape `(output_depth, output_height, output_width, 3)`.
- Grid values are normalized `align_corners=False` pull coordinates in `(x, y, z)` order. Values outside `[-1, 1]` are valid and use the selected border mode.
- The output spatial shape comes from `sampling_grid`. The output preserves the input volume's container, dtype, channel count, and public channel layout.
- `cv2.INTER_LINEAR` uses trilinear sampling. `cv2.INTER_NEAREST` uses nearest-neighbor sampling.
- `cv2.BORDER_CONSTANT` and `cv2.BORDER_REPLICATE` are the initial border modes.
- Constant fill accepts a scalar or one value per channel. Nonzero fill follows the existing subtract-fill, sample, add-fill approach in `warp_affine3d`.
- `uint8` performs one float32 working conversion and one final saturating round-to-`uint8` conversion.
- The primitive runs under `torch.no_grad()` so its Tensor output can be passed to a later trainable module.
- The caller validates rank, layout, dtype, device, parameter values, and autograd state before dispatch.

`remap3d` must not scan the dense grid for identity, normalize a voxel-coordinate map, or create a second full grid. The caller owns zero-work identity detection and normalized-grid construction.

## Route decision

The NumPy `DHWC -> DHWC` route is required. It may use zero-copy NumPy/Torch views and CPU `grid_sample` internally.

Prototype both direct Tensor data routes:

1. Tensor `CDHW` volume with a NumPy sampling grid.
2. Tensor `CDHW` volume with a Tensor sampling grid.

Compare each complete Tensor-to-Tensor call with `Tensor -> NumPy -> remap3d NumPy route -> Tensor` on pre-created, logically identical inputs. Retain a public Tensor overload only when the fastest correctness-equivalent direct route has no repeatable slowdown in any required benchmark cell. Treat results within calibrated timing noise, which must be at most 1%, as a tie and prefer the direct route. One repeatable losing cell rejects the public Tensor overload for the first release; do not add size, dtype, channel, stride, or shape thresholds.

Delete rejected production candidates after retaining their raw benchmark results and a short decision record.

## Correctness matrix

Add an independent scalar nearest/trilinear reference and cover:

- identity grids, coordinate ramps, axis impulses, integer translations, and half-voxel samples;
- non-cubic output shapes and unit depth, height, or width;
- constant zero, nonzero scalar, nonzero per-channel, and replicate borders;
- `uint8` and `float32` volumes with 1, 3, 5, and one 9-channel case;
- NumPy/Tensor parity after layout conversion for every accepted route;
- positive-stride views, negative-stride NumPy inputs, read-only NumPy inputs, and non-contiguous Tensor inputs;
- no mutation of caller-owned volume or grid storage;
- a returned Tensor feeding a downstream trainable module.

The public primitive must call `grid_sample` exactly once per volume.

## Performance and memory matrix

Benchmark the complete public call on the same stable CPU host with fixed thread counts and pre-created inputs. The required matrix is:

| Axis | Values |
| --- | --- |
| Shape | `32^3`, `64 x 128 x 128`, `128^3` |
| Channels | 1, 3 |
| Dtype | `uint8`, `float32` |
| Volume container | NumPy `DHWC`, direct Tensor `CDHW` candidate |
| Grid container | NumPy, Tensor where applicable |
| Tensor strides | contiguous `CDHW`, channel-last-derived strided `CDHW` |
| Interpolation | nearest, linear |
| Border | constant zero, nonzero fill, replicate |

Record hardware, Python/NumPy/Torch/OpenCV versions, thread settings, warmup, raw repetitions, median, MAD or IQR, input strides, and peak RSS. Use one thread for the route gate and report a fixed physical-core count only as a diagnostic.

Peak-memory measurements must show no second `(D, H, W, 3)` grid and no defensive full-volume clone for supported inputs. Report unavoidable output and `uint8` float32 working allocations separately.

If shared-helper extraction changes `warp_affine3d`, run its full correctness and performance matrix on the same exact head. Do not accept an unexplained regression above 5% in an existing `warp_affine3d` route.

## Expected files

- `albucore/remap3d.py`: public router, CPU Torch backend, dtype/fill/container restoration.
- `albucore/geometric.py`, `albucore/__init__.py`, and `docs/public-api.md`: router export and documented public surface.
- `tests/test_remap3d.py`: focused examples and independent reference.
- `tests/property/test_remap3d_properties.py`: dtype, layout, range, shape, and container properties.
- `benchmarks/benchmark_remap3d.py`: route, stride, dtype, border, and peak-memory evidence.
- A small private sampling helper only if sharing it keeps both `remap3d` and `warp_affine3d` fast.

## Non-goals

- Elastic-field or control-grid generation.
- Full-resolution random noise or 3D Gaussian smoothing.
- Voxel-to-normalized coordinate conversion inside `remap3d`.
- GPU, CUDA, MPS, autograd through the primitive, `torch.compile`, or batched public volumes.
- Physical-spacing-aware deformation.
- A runtime backend selector without measured winning regions.

## Acceptance criteria

- The normalized pull-grid contract is documented and covered by an independent reference.
- The NumPy route preserves `DHWC`, dtype, channels, border behavior, and fill behavior.
- The direct Tensor overload is either supported by the complete route matrix or absent; no losing direct route remains behind a heuristic router.
- Each call performs one volumetric resampling operation and does not materialize a second dense grid.
- Peak RSS, raw latency results, and rejected candidates are attached to the implementation PR.
- `warp_affine3d` behavior remains unchanged, with no unexplained full-route regression above 5% if its internals are refactored.
- Focused tests, property tests, type checks, Ruff, and pre-commit pass on the exact implementation head.

**Comment by ternaus** [comment]
Implemented.
